### PR TITLE
Extending LFortran with requirement and template

### PR DESCRIPTION
**Pull request by ansharlubis**
Does not yet produce LLVM code properly.

### CODE REVIEW

**Comment by certik** [comment]
@ansharlubis let me know if you want me to review.

**Comment by certik** [comment]
Yes, I actually reviewed it today, the ASR looks good. I'll write more soon.

**Comment by certik** [comment]
Ok, I just merged #941. There are some conflicts here, so we have to resolve them.

**Comment by certik** [comment]
Ok, so the ASR looks good. I pushed in some changes to resolve some conflicts and polished the parser.

I think the `type_params` are not filled in for `Function`? We can fix that in subsequent PRs. The LLVM seems to be correct?

It seems to run:
```
$ lfortran integration_tests/template_add.f90 
The result is  1.22999992e+01
The result is  14
```

Let's polish this PR and merge. Then we can do subsequent PRs with further improvements.

**Comment by certik** [comment]
Currently we have the following in `Function`:

`ttype* type_params, symbol* restrictions`

And we can (for now) require that `add_t(T, F)` has to specify all the arguments in exactly the following order: first all templates, then all functions interfaces (restrictions/requirements), and we store all templates in `type_params` and all subsequent functions in `restrictions`. Then at instantiation time, we get something like `instantiate add_t(real, func_arg_real)`, so we identify the arguments exactly in the same order, first all `type_params`, then all `restrictions`. This is not the most general way but it will get us started.

Later, we can consider making `FunctionType` a type, see https://github.com/lfortran/lfortran/pull/960, then we can mix templates and functions in any order, as we would store all of them in `type_params` (we would remove `restrictions`).

**Comment by certik** [comment]
Another issue is how to represent the template/requirement structure, something like:
```diff
--- a/src/libasr/ASR.asdl
+++ b/src/libasr/ASR.asdl
@@ -88,6 +88,10 @@ symbol
         expr? return_var, abi abi, access access, deftype deftype,
         string? bindc_name, bool elemental, bool pure, bool module, bool inline,
         bool static, ttype* type_params, symbol* restrictions, bool is_restriction)
+    | Requirement(ttype* type_params, symbol* restrictions)
+    | Template(ttype* type_params, symbol* restrictions,
+            -- plus instantiation of requirements...
+            symbol* requirements, symbol* functions)
     | GenericProcedure(symbol_table parent_symtab, identifier name,
         symbol* procs, access access)
     | CustomOperator(symbol_table parent_symtab, identifier name,
```
But it seems overly complex. A better way would be just use functions.

**Comment by ansharlubis** [comment]
@certik I got the implementation cleaned up with the new AST. It also passed the tests.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -779,6 +785,14 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
                 }
             } else {
+                // DONE: fix this ad-hoc solution ==> remove this solution
+                /*
+                if (is_instantiate) {
+                    ASR::ttype_t *type = LFortran::ASRUtils::TYPE(ASR::make_Character_t(al, loc,
+                            1, strlen(s2c(al, var_name)), nullptr, nullptr, 0));
+                    return ASR::make_StringConstant_t(al, loc, s2c(al, var_name), type);
+                }
+                */
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:795]:
```suggestion
```

**Review by certik — APPROVED** [review]
I think that this looks great. Let's merge this, and we can improve upon this with subsequent PRs. Thanks for finishing this!

**Pull request merged**